compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmOptimizationLowering.kt

```diff
@@ -63,6 +65,42 @@ class JvmOptimizationLowering(val context: JvmBackendContext) : FileLoweringPass
             else -> null
         }
 
+    private fun IrExpression.coerceTo(target: IrType): IrExpression =
+        IrCallImpl(
+            startOffset,
+            endOffset,
+            target,
+            context.ir.symbols.unsafeCoerceIntrinsic
+        ).also { call ->
+            call.putTypeArgument(0, type)
+            call.putTypeArgument(1, target)
+            call.putValueArgument(0, this)
+        }
+
+    // Assumes both original and replacement are static functions, hence does not copy receivers.
+    private fun IrCall.replaceWithCallTo(replacement: IrFunctionSymbol) =
+        IrCallImpl(
+            startOffset,
+            endOffset,
+            type,
+            replacement
+        ).also { newCall ->
+            (0 until valueArgumentsCount).forEach {
+                newCall.putValueArgument(it, getValueArgument(it)!!.coerceTo(replacement.owner.valueParameters[it].type))
+            }
+        }
+
+    private val Jvm8builtInReplacements = mapOf(
+        ("kotlin.UInt" to "compareTo-WZ4Q5Ns") to context.ir.symbols.compareUnsignedInt,
+        ("kotlin.UInt" to "div-WZ4Q5Ns") to context.ir.symbols.divideUnsignedInt,
+        ("kotlin.UInt" to "rem-WZ4Q5Ns") to context.ir.symbols.remainderUnsignedInt,
+        ("kotlin.UInt" to "toString-impl") to context.ir.symbols.toUnsignedStringInt,
+        ("kotlin.ULong" to "compareTo-VKZWuLQ") to context.ir.symbols.compareUnsignedLong,
+        ("kotlin.ULong" to "div-VKZWuLQ") to context.ir.symbols.divideUnsignedLong,
+        ("kotlin.ULong" to "rem-VKZWuLQ") to context.ir.symbols.remainderUnsignedLong,
+        ("kotlin.ULong" to "toString-impl") to context.ir.symbols.toUnsignedStringLong
+    )
```

Comment:
Can we invoke the name mangling methods here, passing the required signature information as strings, to get strings such as `-WZ4Q5Ns` back without manually hardcoding them here? So that it would be less painful if we change the name mangling scheme once again in the future.

Reply:
NB it might require some refactoring to isolate common functionality from inlineClassManglingUtils.kt, where it is implemented in terms of Descriptors and KotlinTypes. It all boils down to building a String from type FQNs and taking base64 MD5.

Reply:
@udalov We could try something like that! But it would require some reshuffling of utility functions as @dnpetrov points out.